design-docs/opmodes.md

```diff
@@ -0,0 +1,516 @@
+
+# Migration from 2025 FRC (WPILib)
+
+Key differences from 2025 FRC:
+- The per-opmode overrideable functions in Robot are replaced with separate annotated per-opmode classes (for non-command-based)
+- `SendableChooser` is no longer required for selecting autonomous routines; instead this functionality is integrated into opmodes, as users can create/register multiple autonomous opmodes classes, and it's been extended to support multiple teleoperated and multiple test opmodes
+- Selection of autonomous opmodes is integrated into the DS instead of being performed by the dashboard
+- OpModes are no longer limited to just timed (periodic); a mix of different opmode functionality/approaches across different robot opmodes is possible (e.g. teleop can be periodic and autonomous linear)
+- Linear opmodes are now available; these are improved versions of the old SimpleRobot/SampleRobot
+
+# Migration from 2025 FTC (FTC SDK)
+
+Key differences from 2025 FTC:
+- There is no hardware map.  Hardware objects are instead directly instantiated inside a top-level Robot class.  This Robot class is provided as a parameter to the opmode constructor.
+- Init is integrated into the opmode constructor.  There is no equivalent to the the enabled init step; the opmode constructor is run as soon as the opmode is selected on the DS, while the robot is still disabled.  Match periods will start as soon as the robot is enabled.
+- There is no `@Disabled` annotation for opmodes; the opmode annotation can simply be commented out to achieve the same result.
+
+# Drawbacks
+
+# Alternatives
+
+Use SendableChooser for more opmodes (teleop and test as well as autonomous); downsides of this:
+- Doesn't allow flexibility of mixing linear and periodic opmodes
+- Harder to use than annotations (generic class)
+- No grouping, no filtering by opmode (although these could be added)
+
+# Trades
+
+- Overall naming: mode? routine? opmode?
+
+- Naming of opmode functions? start-periodic-end, vs init-periodic-exit (2025 FRC PeriodicRobot), vs init-execute-end (2025 FRC Command), vs init-start-loop-stop (2025 FTC OpMode; note init behaves like the constructor here)
+
+- For matches, should we construct teleop at the same time as auto?  If we do that, we probably need a disabledStart() or 2025 FTC opmode style init() that's run on teleop after auto completes, and don't run disabledPeriodic for both simultaneously.
+
+# Unresolved Questions
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousOpMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
```

Comment:
Annotation processing could detect at compile-time, right?

Reply:
We could do that, yes.  I just would want to avoid doing any compile time code generation; we’ve found that to not play nicely with Gradle/IDE support.

Reply:
Would a compile time code checker be feasible though? I'd guess that we'd run into less issues since we wouldn't need to consume outputs of the annotation processing.